dev-docs/yaml-design-guidance.md

```diff
@@ -0,0 +1,293 @@
+Every container that takes part in config should be annotated with `#[serde(deny_unknown_fields)]`. If not the user can make mistakes on their config and they they won't get errors.
+
+#### GOOD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url
+}
+```
+```yaml
+export: 
+  url: http://example.com
+  backup: http://example2.com # The user will receive an error for this
+```
+
+#### BAD
+```rust
+struct Export {
+    url: Url
+}
+```
+```yaml
+export: 
+  url: http://example.com
+  backup: http://example2.com # The user will NOT receive an error for this
+```
+
+### Don't use `#[serde(flatten)]`
+Serde flatten is tempting to use where you have identified common functionality, but creates a bad user experience as it is incompatible with `#[serde(deny_unknown_fields)]`. There isn't a great solution to this, but nesting config can sometimes help.
+
+See [serde documentation](https://serde.rs/field-attrs.html#flatten) for more details.
+
+#### MAYBE
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url,
+    backup: Url
+}
+#[serde(deny_unknown_fields)]
+struct Telemetry {
+    export: Export
+}
+#[serde(deny_unknown_fields)]
+struct Metrics {
+    export: Export
+}
+```
+```yaml
+telemetry:
+  export: 
+    url: http://example.com
+    backup: http://example2.com
+metrics:
+  export:
+    url: http://example.com
+    backup: http://example2.com
+```
+
+#### BAD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url,
+    backup: Url
+}
+struct Telemetry {
+    export: Export
+}
+```
+```yaml
+telemetry: 
+  url: http://example.com
+  backup: http://example2.com
+  unknown: sadness # The user will NOT receive an error for this
+```
+
+### Use consistent terminology
+Be consistent with the rust API terminology.
+* request - functionality that modifies the request or retrieves data from the request of a service.
+* response - functionality that modifies the response or retrieves data from the response of a service.
+* supergraph - functionality within Plugin::subgraph_service
+* execution - functionality within Plugin::execution_service
+* subgraph(s) - functionality within Plugin::subgraph_service
+
+If you use the above terminology then changes are you are doing something that will take place on every request. In this case make sure to include an `action` verb so the user know what the config is doing.
+
+#### GOOD
+```yaml
+headers:
+  subgraphs: # Modifies the subgraph service  
+    products: 
+      request: # Retrieves data from the request
+        - propagate: # The action.
+            named: foo
+```
+
+#### BAD
+```yaml
+headers:
+  named: foo # From where, what are we doing, when is it happening?
+```
+
+### Document your configuration options
+If your config is well documented in Rust then it will be well documented in the generated Json Schema. This means that when users are modifying their config either in their IDE or in Apollo GraphOS documentation is available.
```

Comment:
```suggestion
If your config is well documented in Rust then it will be well documented in the generated JSON Schema. This means that when users are modifying their config either in their IDE or in Apollo GraphOS, documentation is available.
```